Generate ExportDataSchema for new exports

benrudolph opened the pull request:
@NoahCarnahan @snopoke @czue 
ref: https://github.com/dimagi/commcare-hq/pull/9985

NoahCarnahan committed:
add tests

NoahCarnahan committed:
Implement TableConfiguration

Commit:
Separate old and new models

Commit:
add Meta app label to old models

Commit:
Abstract get_built_app_ids_from_app_id to dbaccessors

Commit:
MVP for parsing XForms

Commit:
Rename models

Commit:
Generate schema for form

Commit:
Remove unused constant

Commit:
merge ExportDataSchemas

Commit:
Remove unused attrs

Commit:
Update doc strings

Commit:
TableConfiguration -> ExportDataSchema

Commit:
Test for generate_schema_from_builds

Commit:
don't use wrapped question

Commit:
[ci skip] dimagimon

Commit:
revert submodule

dimagimon commented:
There are 18 errors:

* corehq/apps/export/models/__init__.py, line 1 - F401 'FormQuestionSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 1 - F401 'QuestionMeta' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportColumn' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportDataSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportGroupSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportInstance' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportRow' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'MultipleChoiceItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'Option' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ScalarItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'TableConfiguration' imported but unused
* corehq/apps/export/models/new.py, line 92 - E501 line too long (129 > 115 characters)
* corehq/apps/export/tests/test_table_configuration.py, line 99 - E303 too many blank lines (2)
* corehq/apps/export/tests/test_table_configuration.py, line 105 - F821 undefined name 'SplitExportColumn'
* corehq/apps/export/tests/test_table_configuration.py, line 106 - F821 undefined name 'MultipleChoiceItem'
* corehq/apps/export/tests/test_table_configuration.py, line 112 - F821 undefined name 'SplitExportColumn'
* corehq/apps/export/tests/test_table_configuration.py, line 113 - F821 undefined name 'MultipleChoiceItem'

NoahCarnahan committed:
Update ExportInstance structure

NoahCarnahan committed:
Rename _get_items_for_repeat to _get_sub_documents

NoahCarnahan committed:
Fix split column test

NoahCarnahan committed:
Implement SplitExportColumn

SplitExportColumns are used to separate each option from a multi
choice question into its own column

dimagimon commented:
There are 14 errors:

* corehq/apps/export/models/__init__.py, line 1 - F401 'FormQuestionSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 1 - F401 'QuestionMeta' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportColumn' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportDataSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportGroupSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportInstance' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportRow' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'MultipleChoiceItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'Option' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ScalarItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'TableConfiguration' imported but unused
* corehq/apps/export/models/new.py, line 96 - E501 line too long (126 > 115 characters)
* corehq/apps/export/tests/test_table_configuration.py, line 101 - E303 too many blank lines (2)

NoahCarnahan committed:
[ci skip] lint

dimagimon commented:
There are 13 errors:

* corehq/apps/export/models/__init__.py, line 1 - F401 'FormQuestionSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 1 - F401 'QuestionMeta' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportColumn' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportDataSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportGroupSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportInstance' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportRow' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'MultipleChoiceItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'Option' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ScalarItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'TableConfiguration' imported but unused
* corehq/apps/export/models/new.py, line 96 - E501 line too long (126 > 115 characters)

NoahCarnahan commented:
@benrudolph I added a few commits to fix original `TableConfiguration` tests

benrudolph commented:
thanks noah. travis is rough today, but looks good barring tests

czue commented on corehq/apps/export/models/new.py:
```diff
@@ -0,0 +1,370 @@
+        app_label = 'export'
+
+
+class ExportRow(object):
```
what's this model for?

NoahCarnahan replied:
I was imagining this class would do something like [`FormattedRow`](https://github.com/dimagi/commcare-hq/blob/04257f5864e3f1d44cbc95e85080369358b868a2/corehq/ex-submodules/couchexport/export.py#L346), but wasn't sure yet what other properties would be needed.

czue commented on corehq/apps/export/models/new.py:
```diff
@@ -0,0 +1,370 @@
+    last_occurrence = IntegerProperty()
+
+
+class ExportDataSchema(DocumentSchema):
```
is this not a `Document`?

benrudolph replied:
it probably will be, but the models are still in flux a bit. there's a possibility this could be part of the ExportConfiguration

czue replied:
does that imply that there wouldn't be a global object? would you rebuild it from the apps every time you need it?

benrudolph replied:
part of me really wants to rebuild it every time. maintaining state is always a nightmare (just thinking if there's a bug in the schema generation, we'd have to have a way to force rebuild... or just create a new export). i think in terms of performance it makes way more sense to save the schema in the global `ExportConfiguration` object

czue commented:
looking good!

NoahCarnahan committed:
Update `SplitExportColumn.get_value()` doc string

dimagimon commented:
There are 13 errors:

* corehq/apps/export/models/__init__.py, line 1 - F401 'FormQuestionSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 1 - F401 'QuestionMeta' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportColumn' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportDataSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportGroupSchema' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportInstance' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ExportRow' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'MultipleChoiceItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'Option' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'ScalarItem' imported but unused
* corehq/apps/export/models/__init__.py, line 2 - F401 'TableConfiguration' imported but unused
* corehq/apps/export/models/new.py, line 96 - E501 line too long (126 > 115 characters)

benrudolph commented:
(noah and I pressed the merge button together)

benrudolph commented:
:two_men_holding_hands:
